### ISSUE TITLE
Mimi Training

### ISSUE THREAD

**Issue by yukiarimo** [senstella/csm-mlx#20]
Hello! Can you please implement Mimi Codec training? I tried, but always getting either nan loss or model doesn't update :(

**Comment by yukiarimo** [comment]
Code to reproduce:

```
import argparse
import time
from pathlib import Path
import librosa
import mlx.core as mx
import mlx.nn as nn
import mlx.optimizers as optim
import numpy as np
from tqdm import tqdm

from moshi_mlx.models.mimi import Mimi, mimi_202407

def load_audio_chunks(audio_dir: Path, chunk_duration: float, sample_rate: int):
    """
    A generator that finds .wav files, loads, resamples, and yields fixed-length chunks.
    """
    print(f"Searching for .wav files in {audio_dir}...")
    # Using rglob to find files in subdirectories as well
    audio_paths = list(audio_dir.rglob("*.wav"))
    if not audio_paths:
        raise FileNotFoundError(f"No .wav files found in {audio_dir}")

    print(f"Found {len(audio_paths)} audio files. Processing into chunks...")
    
    chunk_samples = int(chunk_duration * sample_rate)
    
    for path in tqdm(audio_paths, desc="Loading and chunking audio"):
        try:
            # Load, resample to the target rate, and convert to mono
            audio, sr = librosa.load(path, sr=sample_rate, mono=True)
            
            # Skip short audio files
            if len(audio) < chunk_samples:
                continue

            # Create chunks and discard the last partial chunk
            for start_frame in range(0, len(audio) - chunk_samples + 1, chunk_samples):
                end_frame = start_frame + chunk_samples
                yield audio[start_frame:end_frame]
        except Exception as e:
            print(f"Warning: Could not process file {path}. Error: {e}")

def loss_fn(model: Mimi, x: mx.array) -> mx.array:
    """
    Computes the reconstruction loss for a batch of audio.
    The loss is the Mean Absolute Error (L1 loss) between the original and reconstructed waveform.
    """
    # Encode and then decode the audio to get the reconstruction
    tokens = model.encode(x)
    reconstructed_x = model.decode(tokens)

    if mx.isnan(reconstructed_x).any():
        print("Warning: NaN detected in reconstructed audio!")

    # The reconstructed audio might have a slightly different length due to convolution padding.
    # We must trim both to the minimum length for a valid comparison.
    original_len = x.shape[-1]
    reconstructed_len = reconstructed_x.shape[-1]
    min_len = min(original_len, reconstructed_len)

    x_trimmed = x[..., :min_len]
    reconstructed_trimmed = reconstructed_x[..., :min_len]

    # Calculate L1 loss
    loss = mx.mean(mx.abs(x_trimmed - reconstructed_trimmed))
    return loss

def main():
    parser = argparse.ArgumentParser(description="Train or fine-tune the Mimi audio codec.")
    parser.add_argument("--audio-dir", type=str, required=True,
                        help="Directory containing .wav files for training.")
    parser.add_argument("--pretrained-model-path", type=str,
                        help="Path to a pre-trained Mimi model (.safetensors) for fine-tuning. If not provided, training starts from scratch.")
    parser.add_argument("--output-dir", type=str, default="mimi_finetuned",
                        help="Directory to save model checkpoints.")
    parser.add_argument("--epochs", type=int, default=10,
                        help="Number of training epochs.")
    parser.add_argument("--learning-rate", type=float, default=1e-4,
                        help="Optimizer learning rate.")
    parser.add_argument("--batch-size", type=int, default=8,
                        help="Number of audio chunks per training batch.")
    parser.add_argument("--num-codebooks", type=int, default=32,
                        help="Number of codebooks in the Mimi model.")
    
    args = parser.parse_args()

    # Setup
    output_dir = Path(args.output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    sample_rate = 48000
    chunk_duration = 4.0

    # --- 1. Load Dataset ---
    all_chunks_generator = load_audio_chunks(Path(args.audio_dir), chunk_duration, sample_rate)
    # Convert generator to a list to easily shuffle and batch
    all_chunks = list(all_chunks_generator)
    if not all_chunks:
        print("Error: No audio chunks could be created from the provided directory. Check your audio files.")
        return
        
    print(f"Successfully created {len(all_chunks)} chunks of {chunk_duration} seconds each.")

    # --- 2. Initialize Model ---
    print("Initializing Mimi model...")
    model = Mimi(mimi_202407(args.num_codebooks))
    model.load_weights("model_path.safetensors")

    for idx, param in enumerate(model.parameters()):
        if isinstance(param, mx.array):
            if mx.isnan(param).any():
                print(f"Warning: NaN detected in parameter at index {idx} after loading weights!")
    
    model.train()
    mx.eval(model.parameters())

    # --- 3. Initialize Optimizer and Loss Function ---
    optimizer = optim.AdamW(learning_rate=args.learning_rate)
    loss_and_grad_fn = nn.value_and_grad(model, loss_fn)

    # --- 4. Training Loop ---
    print("\nStarting training...")
    for epoch in range(1, args.epochs + 1):
        epoch_start_time = time.time()
        total_loss = 0.0
        num_batches = 0
        
        # Shuffle chunks for each epoch
        np.random.shuffle(all_chunks)

        # Create batches
        pbar = tqdm(range(0, len(all_chunks), args.batch_size), desc=f"Epoch {epoch}/{args.epochs}")
        for i in pbar:
            batch_np = all_chunks[i : i + args.batch_size]
            if not batch_np: # Skip last empty batch
                continue
            
            # Convert batch to MLX array: (batch, channels, time)
            batch_mx = mx.array(np.stack(batch_np))[:, None, :]

            # Compute loss and gradients
            loss, grads = loss_and_grad_fn(model, batch_mx)

            # Update model weights
            optimizer.update(model, grads)
            
            # This is crucial to execute the computation graph
            mx.eval(model.parameters(), optimizer.state)

            batch_loss = loss.item()
            total_loss += batch_loss
            num_batches += 1
            pbar.set_postfix({"loss": f"{batch_loss:.6f}"})

        # --- End of Epoch ---
        avg_loss = total_loss / num_batches if num_batches > 0 else 0
        epoch_time = time.time() - epoch_start_time
        print(f"Epoch {epoch} finished in {epoch_time:.2f}s. Average Loss: {avg_loss:.6f}")

        # --- 5. Checkpointing ---
        checkpoint_path = output_dir / f"mimi_checkpoint_epoch_{epoch}.safetensors"
        model.save_weights(str(checkpoint_path))
        print(f"Saved checkpoint to {checkpoint_path}")

    print("\nTraining complete!")
    print(f"Final fine-tuned models are saved in: {output_dir}")

if __name__ == "__main__":
    main()
```

**Comment by senstella** [comment]
Hello! Thanks for the suggestion. However, the `csm-mlx` library is specifically for the CSM model, and I actually import the MLX Mimi implementation from the [moshi](https://github.com/kyutai-labs/moshi) repository rather than implementing it myself. The moshi repo would probably be the best place to ask about this type of implementation!
